Add lookup for `LocalRootSpanId` property and tests for `SpanId` and `LocalRootSpanId`.

macrogreg opened the pull request:
A [recent PR](https://github.com/DataDog/dd-trace-dotnet/pull/1831) suggested to add direct access properties to Span to access the id and the id of the root span of the local non-reentrant sub-trace. That PR added private fields to `Span` to store those values, to avoid dereferencing multiple pointers at every lookup. However, it was not clear, whether that memory increase was justified, however small it was.

To address that, _this_ PR provides and tests those properties without using additional memory. We are adding using tests, and will keep https://github.com/DataDog/dd-trace-dotnet/pull/1831 archived in case we need to get back to using private fields.

macrogreg commented:
@kevingosse you suggested elsewhere to name the new property `LocalRootSpanId` rather than `RootSpanId`. I do not mind which of the too, so happy to change it, but let me first provide context into why I named it `LocalRootSpanId`:

In short:

To avoid confusion between the global trace root and the local sub-trace root in the context of distributed and re-entrant traces.

The unit tests, for example, show a case where the _local root span_ has a parent id, but it represents the _remote_ (aka propagated) parent, so does not play a role in determining the local root. So I picked a name that makes it explicitly clear.

In long:

A _trace_ is an operation distributed across many microservices (or, most generically, processes). A _span_ can represent any logical sub-operation that does not cross the process-boundary. Typically, but not always, the local sub-operation is represented by several hierarchically organized spans.

A trace has a _root span_ that represents the "first" (aka the outer-most) sub-operation of the entire distributed operation represented by that trace. There is exactly one root span in each distributed trace. (A common implementation detail is that the root span is the one that also initializes the entire trace context, including the trace id, which is then propagated across process boundaries.)

Each sub-operation executed by a participating service can have many spans and will have exactly one local root span. Notably, traces can be reentrant, e.g. service-A calls service-B which, in turn, calls service-A. So some service-A may concurrently have more than one execution stack and more than one local root span belonging to the same distributed trace.

All that rarely matters for local data collection, but matters for piecing together distributed trace info or for relating data from different pipelines, which is what Code Hotspots is doing.

andrewlock commented:
## Benchmarks Report :snail:

Benchmarks for #1839 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81): 
* All benchmarks have **the same** speed
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`WriteAndFlushEnrichedTraces`|net472|897.3&nbsp;μs|15.67&nbsp;μs|14.66&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|695.8&nbsp;μs|5.69&nbsp;μs|5.32&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
|#1839|`WriteAndFlushEnrichedTraces`|net472|912.8&nbsp;μs|16.86&nbsp;μs|15.78&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|#1839|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|721.9&nbsp;μs|14.42&nbsp;μs|14.16&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendRequest`|netcoreapp3.1|259,157.8237&nbsp;ns|5,006.2051&nbsp;ns|7,493.0521&nbsp;ns|0.2525| 0.0000|0.0000|18664&nbsp;B|
|#1839|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|#1839|`SendRequest`|netcoreapp3.1|257,149.1491&nbsp;ns|5,038.8192&nbsp;ns|7,063.7208&nbsp;ns|0.2456| 0.0000|0.0000|18664&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`ExecuteNonQuery`|net472|1.946&nbsp;μs|0.0380&nbsp;μs|0.0407&nbsp;μs|0.1077| 0.0009|0.0000|682&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`ExecuteNonQuery`|netcoreapp3.1|1.827&nbsp;μs|0.0249&nbsp;μs|0.0208&nbsp;μs|0.0100| 0.0000|0.0000|712&nbsp;B|
|#1839|`ExecuteNonQuery`|net472|2.005&nbsp;μs|0.0388&nbsp;μs|0.0363&nbsp;μs|0.1079| 0.0010|0.0000|682&nbsp;B|
|#1839|`ExecuteNonQuery`|netcoreapp3.1|1.897&nbsp;μs|0.0308&nbsp;μs|0.0273&nbsp;μs|0.0103| 0.0000|0.0000|712&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`CallElasticsearch`|net472|2.874&nbsp;μs|0.0557&nbsp;μs|0.0572&nbsp;μs|0.1279| 0.0000|0.0000|827&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`CallElasticsearch`|netcoreapp3.1|1.650&nbsp;μs|0.0156&nbsp;μs|0.0146&nbsp;μs|0.0104| 0.0000|0.0000|768&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`CallElasticsearchAsync`|net472|3.065&nbsp;μs|0.0371&nbsp;μs|0.0329&nbsp;μs|0.1498| 0.0000|0.0000|963&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`CallElasticsearchAsync`|netcoreapp3.1|1.609&nbsp;μs|0.0167&nbsp;μs|0.0156&nbsp;μs|0.0118| 0.0000|0.0000|888&nbsp;B|
|#1839|`CallElasticsearch`|net472|2.847&nbsp;μs|0.0304&nbsp;μs|0.0284&nbsp;μs|0.1284| 0.0000|0.0000|827&nbsp;B|
|#1839|`CallElasticsearch`|netcoreapp3.1|1.517&nbsp;μs|0.0134&nbsp;μs|0.0125&nbsp;μs|0.0109| 0.0000|0.0000|768&nbsp;B|
|#1839|`CallElasticsearchAsync`|net472|3.041&nbsp;μs|0.0366&nbsp;μs|0.0342&nbsp;μs|0.1504| 0.0000|0.0000|963&nbsp;B|
|#1839|`CallElasticsearchAsync`|netcoreapp3.1|1.656&nbsp;μs|0.0152&nbsp;μs|0.0134&nbsp;μs|0.0124| 0.0000|0.0000|888&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`ExecuteAsync`|net472|3.275&nbsp;μs|0.0411&nbsp;μs|0.0384&nbsp;μs|0.1684| 0.0000|0.0000|1083&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`ExecuteAsync`|netcoreapp3.1|1.810&nbsp;μs|0.0214&nbsp;μs|0.0200&nbsp;μs|0.0140| 0.0000|0.0000|1008&nbsp;B|
|#1839|`ExecuteAsync`|net472|3.189&nbsp;μs|0.0429&nbsp;μs|0.0402&nbsp;μs|0.1688| 0.0000|0.0000|1083&nbsp;B|
|#1839|`ExecuteAsync`|netcoreapp3.1|1.843&nbsp;μs|0.0252&nbsp;μs|0.0223&nbsp;μs|0.0145| 0.0000|0.0000|1008&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendAsync`|net472|5.664&nbsp;μs|0.0953&nbsp;μs|0.0891&nbsp;μs|0.3502| 0.0000|0.0000|2.19&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendAsync`|netcoreapp3.1|4.088&nbsp;μs|0.0697&nbsp;μs|0.0652&nbsp;μs|0.0300| 0.0000|0.0000|2.09&nbsp;KB|
|#1839|`SendAsync`|net472|5.647&nbsp;μs|0.0590&nbsp;μs|0.0552&nbsp;μs|0.3491| 0.0000|0.0000|2.19&nbsp;KB|
|#1839|`SendAsync`|netcoreapp3.1|4.064&nbsp;μs|0.0274&nbsp;μs|0.0229&nbsp;μs|0.0290| 0.0000|0.0000|2.09&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|net472|3.679&nbsp;μs|0.0718&nbsp;μs|0.0705&nbsp;μs|0.2865| 0.0000|0.0000|1.79&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|netcoreapp3.1|3.503&nbsp;μs|0.0686&nbsp;μs|0.0641&nbsp;μs|0.0275| 0.0000|0.0000|1.94&nbsp;KB|
|#1839|`EnrichedLog`|net472|3.652&nbsp;μs|0.0522&nbsp;μs|0.0436&nbsp;μs|0.2872| 0.0000|0.0000|1.79&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|3.590&nbsp;μs|0.0465&nbsp;μs|0.0435&nbsp;μs|0.0268| 0.0000|0.0000|1.94&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|net472|171.5&nbsp;μs|3.33&nbsp;μs|3.12&nbsp;μs|1.3186| 0.3516|0.0000|8.95&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|netcoreapp3.1|153.4&nbsp;μs|2.84&nbsp;μs|2.79&nbsp;μs|0.0758| 0.0000|0.0000|9.82&nbsp;KB|
|#1839|`EnrichedLog`|net472|169.5&nbsp;μs|1.99&nbsp;μs|1.86&nbsp;μs|1.3003| 0.3467|0.0000|8.95&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|145.8&nbsp;μs|1.41&nbsp;μs|1.32&nbsp;μs|0.0761| 0.0000|0.0000|9.82&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|net472|10.955&nbsp;μs|0.2009&nbsp;μs|0.1879&nbsp;μs|0.8524| 0.0000|0.0000|5.31&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|netcoreapp3.1|9.611&nbsp;μs|0.1730&nbsp;μs|0.1618&nbsp;μs|0.0900| 0.0000|0.0000|6.28&nbsp;KB|
|#1839|`EnrichedLog`|net472|10.872&nbsp;μs|0.1352&nbsp;μs|0.1265&nbsp;μs|0.8557| 0.0000|0.0000|5.31&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|9.574&nbsp;μs|0.1301&nbsp;μs|0.1153&nbsp;μs|0.0865| 0.0000|0.0000|6.28&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendReceive`|net472|2.054&nbsp;μs|0.0386&nbsp;μs|0.0397&nbsp;μs|0.1549| 0.0000|0.0000|987&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`SendReceive`|netcoreapp3.1|1.839&nbsp;μs|0.0187&nbsp;μs|0.0156&nbsp;μs|0.0137| 0.0000|0.0000|984&nbsp;B|
|#1839|`SendReceive`|net472|2.039&nbsp;μs|0.0253&nbsp;μs|0.0211&nbsp;μs|0.1543| 0.0000|0.0000|987&nbsp;B|
|#1839|`SendReceive`|netcoreapp3.1|1.836&nbsp;μs|0.0167&nbsp;μs|0.0156&nbsp;μs|0.0139| 0.0000|0.0000|984&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|net472|7.410&nbsp;μs|0.1163&nbsp;μs|0.1031&nbsp;μs|0.4481| 0.0000|0.0000|2.8&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`EnrichedLog`|netcoreapp3.1|6.786&nbsp;μs|0.1070&nbsp;μs|0.1001&nbsp;μs|0.0376| 0.0000|0.0000|2.61&nbsp;KB|
|#1839|`EnrichedLog`|net472|7.350&nbsp;μs|0.1044&nbsp;μs|0.0977&nbsp;μs|0.4495| 0.0000|0.0000|2.8&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|6.831&nbsp;μs|0.0613&nbsp;μs|0.0543&nbsp;μs|0.0373| 0.0000|0.0000|2.61&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`StartFinishSpan`|net472|844.3&nbsp;ns|7.08&nbsp;ns|6.28&nbsp;ns|0.0679| 0.0000|0.0000|433&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`StartFinishSpan`|netcoreapp3.1|822.6&nbsp;ns|9.42&nbsp;ns|8.81&nbsp;ns|0.0061| 0.0000|0.0000|432&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`StartFinishScope`|net472|1,021.2&nbsp;ns|19.68&nbsp;ns|18.41&nbsp;ns|0.0806| 0.0000|0.0000|514&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/c55f8087c6cc9c8125962f4faa2fc7054d8bee81)|`StartFinishScope`|netcoreapp3.1|995.9&nbsp;ns|9.87&nbsp;ns|9.23&nbsp;ns|0.0074| 0.0000|0.0000|552&nbsp;B|
|#1839|`StartFinishSpan`|net472|846.1&nbsp;ns|16.86&nbsp;ns|15.77&nbsp;ns|0.0682| 0.0000|0.0000|433&nbsp;B|
|#1839|`StartFinishSpan`|netcoreapp3.1|828.0&nbsp;ns|7.88&nbsp;ns|6.98&nbsp;ns|0.0058| 0.0000|0.0000|432&nbsp;B|
|#1839|`StartFinishScope`|net472|1,110.2&nbsp;ns|17.54&nbsp;ns|16.41&nbsp;ns|0.0803| 0.0000|0.0000|514&nbsp;B|
|#1839|`StartFinishScope`|netcoreapp3.1|1,019.2&nbsp;ns|10.96&nbsp;ns|9.15&nbsp;ns|0.0077| 0.0000|0.0000|552&nbsp;B|
</details>

kevingosse commented:
Your argument could have been valid, but I believe this ship has already sailed. The notion of `Root` is already used at multiple places in the code: `Span.IsRootSpan`, `TraceContext.RootSpan`, `Scope.Root`. Introducing the notion of `LocalRoot` only for this one property gives the impression that it's a different concept and only fuels confusion.

macrogreg commented:
> Your argument could have been valid, but I believe this ship has already sailed. The notion of `Root` is already used at multiple places in the code: `Span.IsRootSpan`, `TraceContext.RootSpan`, `Scope.Root`. Introducing the notion of `LocalRoot` only for this one property gives the impression that it's a different concept and only fuels confusion.

Hi @kevingosse,

Fair enough, I updated the PR accordingly.

Btw, when I was working on it, I was quite confused which property is returning what. With @zacharycmontoya 's friendly help we figured it out, but it took several hours. I added a detailed comment to the new property to at least mitigate it for the stuff being added. I realize that ideally we would do it for more APIs properties, but I would like to time-box this effort. Please let me know if this works from your perspective. Thanks!

andrewlock commented:
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/e3341b46eab872744a4569dd979cdb3866c5de33) will **not change** line coverage
:heavy_check_mark: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/e3341b46eab872744a4569dd979cdb3866c5de33) will **not change** branch coverage
:no_entry: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/e3341b46eab872744a4569dd979cdb3866c5de33) will will **increase** complexity by `6`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/e3341b46eab872744a4569dd979cdb3866c5de33) | #1839       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9482` / `15191` | `9477` / `15193` |          |
| Lines %   | `62%`      | `62%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4155` / `6590` | `4156` / `6596` |          |
| Branches %| `63%`      | `63%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7764`      | `7770`        |  `6`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76918&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76918/artifacts?artifactName=Code%20Coverage%20Report_76918&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1839](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76948&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76948/artifacts?artifactName=Code%20Coverage%20Report_76948&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/e3341b46eab872744a4569dd979cdb3866c5de33) | #1839       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `62%`      | `62%`       |  `0%` :heavy_check_mark:  |
| Branches %| `63%`      | `63%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7764`      | `7770`       |  `6` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.Helpers.AsyncHelper](https://github.com/DataDog/dd-trace-dotnet/tree/9a36a2dba8fe43c14c62c95f0cb357b941e4cadc/tracer/src/Datadog.Trace/ClrProfiler/Helpers/AsyncHelper.cs) | `-2%` :warning: | `6%` :heavy_check_mark:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76918&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76918/artifacts?artifactName=Code%20Coverage%20Report_76918&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1839](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76948&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76948/artifacts?artifactName=Code%20Coverage%20Report_76948&api-version=6.0&%24format=zip)

zacharycmontoya commented on tracer/src/Datadog.Trace/Span.cs:
```diff
@@ -85,6 +85,24 @@ public string ServiceName
+        /// <para>If the trace has been propagated from a remote service, the <i>remote global root</i> is not relevant for this API.</para>
+        /// <para>A distributed operation represented by a trace may be re-entrant (e.g. service-A calls service-B, which calls service-A again).
+        /// In such cases, the local process may be concurrently executing multiple local root spans.
+        /// This API returns the if of the root span of the non-reentrant trace sub-set.</para></remarks>
```
```suggestion
        /// This API returns the id of the root span of the non-reentrant trace sub-set.</para></remarks>
```

macrogreg replied:
Thanks!

✔️

zacharycmontoya commented on tracer/test/Datadog.Trace.Tests/SpanTests.cs:
```diff
@@ -160,5 +160,114 @@ public void TopLevelSpans()
+
+        [Fact]
+        public void SpanIds()
+        {
```
My preference would be to break out each scope into its own unit test to ensure that the `_tracer` accesses for each test are independent of each other, since each test case will initiate the constructor

macrogreg replied:
Good point. Done.

✔️

zacharycmontoya reviewed (approved):
LGTM with a couple non-blocking comments

gleocadie commented on tracer/test/Datadog.Trace.Tests/SpanTests.cs:
```diff
@@ -160,5 +160,114 @@ public void TopLevelSpans()
         }
+
+        [Fact]
+        public void SpanIds()
```
I think each block should be a different unit test:
- Easier to debug
- Easier to read/maintain
- Easier to get specific error

macrogreg replied:
Done.

✔️ 

I noticed you blocked the PR. Is there any specific reason?

gleocadie replied:
oops, wrong button :)

andrewlock commented:
## Benchmarks Report :snail:

Benchmarks for #1839 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e): 
* 20 benchmarks are **faster**, with geometric mean 1.157
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`WriteAndFlushEnrichedTraces`|net472|1,016.0&nbsp;μs|12.68&nbsp;μs|10.59&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|790.0&nbsp;μs|15.15&nbsp;μs|18.04&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
|#1839|`WriteAndFlushEnrichedTraces`|net472|908.5&nbsp;μs|14.27&nbsp;μs|14.02&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|#1839|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|714.8&nbsp;μs|14.03&nbsp;μs|13.12&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.AspNetCoreBenchmark.SendRequest&#8209;netcoreapp3.1 | 1.139 | 330,339.39 | 290,088.05 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendRequest`|netcoreapp3.1|331,409.0305&nbsp;ns|6,307.8584&nbsp;ns|6,195.1604&nbsp;ns|0.3281| 0.1640|0.0000|27572&nbsp;B|
|#1839|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|#1839|`SendRequest`|netcoreapp3.1|291,552.2868&nbsp;ns|5,774.2918&nbsp;ns|9,324.4050&nbsp;ns|0.2927| 0.0000|0.0000|27568&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.DbCommandBenchmark.ExecuteNonQuery&#8209;net472 | 1.217 | 2,285.73 | 1,878.07 | 
| Benchmarks.Trace.DbCommandBenchmark.ExecuteNonQuery&#8209;netcoreapp3.1 | 1.206 | 2,188.17 | 1,814.46 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`ExecuteNonQuery`|net472|2.284&nbsp;μs|0.0454&nbsp;μs|0.0666&nbsp;μs|0.1083| 0.0011|0.0000|682&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`ExecuteNonQuery`|netcoreapp3.1|2.191&nbsp;μs|0.0213&nbsp;μs|0.0188&nbsp;μs|0.0090| 0.0000|0.0000|712&nbsp;B|
|#1839|`ExecuteNonQuery`|net472|1.880&nbsp;μs|0.0371&nbsp;μs|0.0364&nbsp;μs|0.1079| 0.0009|0.0000|682&nbsp;B|
|#1839|`ExecuteNonQuery`|netcoreapp3.1|1.818&nbsp;μs|0.0356&nbsp;μs|0.0366&nbsp;μs|0.0097| 0.0000|0.0000|712&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.ElasticsearchBenchmark.CallElasticsearch&#8209;net472 | 1.188 | 3,398.08 | 2,860.70 | 
| Benchmarks.Trace.ElasticsearchBenchmark.CallElasticsearchAsync&#8209;net472 | 1.178 | 3,705.84 | 3,144.62 | 
| Benchmarks.Trace.ElasticsearchBenchmark.CallElasticsearch&#8209;netcoreapp3.1 | 1.126 | 1,716.85 | 1,524.99 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`CallElasticsearch`|net472|3.399&nbsp;μs|0.0457&nbsp;μs|0.0405&nbsp;μs|0.1273| 0.0000|0.0000|827&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`CallElasticsearch`|netcoreapp3.1|1.723&nbsp;μs|0.0204&nbsp;μs|0.0191&nbsp;μs|0.0104| 0.0000|0.0000|768&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`CallElasticsearchAsync`|net472|3.721&nbsp;μs|0.0406&nbsp;μs|0.0380&nbsp;μs|0.1501| 0.0000|0.0000|963&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`CallElasticsearchAsync`|netcoreapp3.1|1.851&nbsp;μs|0.0208&nbsp;μs|0.0174&nbsp;μs|0.0126| 0.0000|0.0000|888&nbsp;B|
|#1839|`CallElasticsearch`|net472|2.856&nbsp;μs|0.0413&nbsp;μs|0.0366&nbsp;μs|0.1292| 0.0000|0.0000|827&nbsp;B|
|#1839|`CallElasticsearch`|netcoreapp3.1|1.524&nbsp;μs|0.0187&nbsp;μs|0.0166&nbsp;μs|0.0107| 0.0000|0.0000|768&nbsp;B|
|#1839|`CallElasticsearchAsync`|net472|3.154&nbsp;μs|0.0555&nbsp;μs|0.0519&nbsp;μs|0.1496| 0.0000|0.0000|963&nbsp;B|
|#1839|`CallElasticsearchAsync`|netcoreapp3.1|1.664&nbsp;μs|0.0145&nbsp;μs|0.0121&nbsp;μs|0.0124| 0.0000|0.0000|888&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.GraphQLBenchmark.ExecuteAsync&#8209;netcoreapp3.1 | 1.129 | 2,061.49 | 1,826.45 | 
| Benchmarks.Trace.GraphQLBenchmark.ExecuteAsync&#8209;net472 | 1.125 | 3,604.27 | 3,202.42 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`ExecuteAsync`|net472|3.619&nbsp;μs|0.0466&nbsp;μs|0.0436&nbsp;μs|0.1688| 0.0000|0.0000|1083&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`ExecuteAsync`|netcoreapp3.1|2.060&nbsp;μs|0.0253&nbsp;μs|0.0236&nbsp;μs|0.0143| 0.0000|0.0000|1008&nbsp;B|
|#1839|`ExecuteAsync`|net472|3.216&nbsp;μs|0.0525&nbsp;μs|0.0439&nbsp;μs|0.1687| 0.0000|0.0000|1083&nbsp;B|
|#1839|`ExecuteAsync`|netcoreapp3.1|1.823&nbsp;μs|0.0182&nbsp;μs|0.0170&nbsp;μs|0.0139| 0.0000|0.0000|1008&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.HttpClientBenchmark.SendAsync&#8209;net472 | 1.134 | 6,400.13 | 5,646.22 | 
| Benchmarks.Trace.HttpClientBenchmark.SendAsync&#8209;netcoreapp3.1 | 1.124 | 4,535.48 | 4,035.08 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendAsync`|net472|6.428&nbsp;μs|0.0956&nbsp;μs|0.0895&nbsp;μs|0.3499| 0.0000|0.0000|2.19&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendAsync`|netcoreapp3.1|4.545&nbsp;μs|0.0478&nbsp;μs|0.0447&nbsp;μs|0.0295| 0.0000|0.0000|2.09&nbsp;KB|
|#1839|`SendAsync`|net472|5.649&nbsp;μs|0.1065&nbsp;μs|0.0996&nbsp;μs|0.3512| 0.0000|0.0000|2.19&nbsp;KB|
|#1839|`SendAsync`|netcoreapp3.1|4.058&nbsp;μs|0.0547&nbsp;μs|0.0511&nbsp;μs|0.0306| 0.0000|0.0000|2.09&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.ILoggerBenchmark.EnrichedLog&#8209;netcoreapp3.1 | 1.262 | 4,439.90 | 3,518.89 | 
| Benchmarks.Trace.ILoggerBenchmark.EnrichedLog&#8209;net472 | 1.149 | 4,171.30 | 3,629.64 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|net472|4.183&nbsp;μs|0.0524&nbsp;μs|0.0464&nbsp;μs|0.2863| 0.0000|0.0000|1.79&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|netcoreapp3.1|4.455&nbsp;μs|0.0446&nbsp;μs|0.0417&nbsp;μs|0.0265| 0.0000|0.0000|1.94&nbsp;KB|
|#1839|`EnrichedLog`|net472|3.635&nbsp;μs|0.0452&nbsp;μs|0.0377&nbsp;μs|0.2863| 0.0000|0.0000|1.79&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|3.524&nbsp;μs|0.0305&nbsp;μs|0.0270&nbsp;μs|0.0266| 0.0000|0.0000|1.94&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Log4netBenchmark.EnrichedLog&#8209;netcoreapp3.1 | 1.130 | 167,546.75 | 148,235.43 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|net472|190.0&nbsp;μs|1.99&nbsp;μs|1.86&nbsp;μs|1.2273| 0.2832|0.0000|8.95&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|netcoreapp3.1|167.5&nbsp;μs|1.69&nbsp;μs|1.58&nbsp;μs|0.0826| 0.0000|0.0000|9.82&nbsp;KB|
|#1839|`EnrichedLog`|net472|170.6&nbsp;μs|2.56&nbsp;μs|2.40&nbsp;μs|1.2755| 0.3401|0.0000|8.95&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|148.5&nbsp;μs|1.70&nbsp;μs|1.50&nbsp;μs|0.0737| 0.0000|0.0000|9.82&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.NLogBenchmark.EnrichedLog&#8209;net472 | 1.131 | 12,520.31 | 11,072.42 | 
| Benchmarks.Trace.NLogBenchmark.EnrichedLog&#8209;netcoreapp3.1 | 1.128 | 10,790.49 | 9,564.73 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|net472|12.50&nbsp;μs|0.162&nbsp;μs|0.143&nbsp;μs|0.8536| 0.0000|0.0000|5.31&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|netcoreapp3.1|10.84&nbsp;μs|0.173&nbsp;μs|0.162&nbsp;μs|0.0865| 0.0000|0.0000|6.28&nbsp;KB|
|#1839|`EnrichedLog`|net472|11.076&nbsp;μs|0.1981&nbsp;μs|0.1853&nbsp;μs|0.8561| 0.0000|0.0000|5.31&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|9.571&nbsp;μs|0.0401&nbsp;μs|0.0313&nbsp;μs|0.0921| 0.0000|0.0000|6.28&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendReceive`|net472|2.298&nbsp;μs|0.0312&nbsp;μs|0.0276&nbsp;μs|0.1548| 0.0000|0.0000|987&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`SendReceive`|netcoreapp3.1|2.107&nbsp;μs|0.0216&nbsp;μs|0.0181&nbsp;μs|0.0135| 0.0000|0.0000|984&nbsp;B|
|#1839|`SendReceive`|net472|2.063&nbsp;μs|0.0363&nbsp;μs|0.0565&nbsp;μs|0.1553| 0.0000|0.0000|987&nbsp;B|
|#1839|`SendReceive`|netcoreapp3.1|1.978&nbsp;μs|0.0378&nbsp;μs|0.0388&nbsp;μs|0.0135| 0.0000|0.0000|984&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SerilogBenchmark.EnrichedLog&#8209;net472 | 1.171 | 8,490.91 | 7,252.28 | 
| Benchmarks.Trace.SerilogBenchmark.EnrichedLog&#8209;netcoreapp3.1 | 1.169 | 7,685.99 | 6,577.23 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|net472|8.484&nbsp;μs|0.1362&nbsp;μs|0.1137&nbsp;μs|0.4485| 0.0000|0.0000|2.8&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`EnrichedLog`|netcoreapp3.1|7.670&nbsp;μs|0.0825&nbsp;μs|0.0772&nbsp;μs|0.0383| 0.0000|0.0000|2.61&nbsp;KB|
|#1839|`EnrichedLog`|net472|7.283&nbsp;μs|0.0984&nbsp;μs|0.0920&nbsp;μs|0.4486| 0.0000|0.0000|2.8&nbsp;KB|
|#1839|`EnrichedLog`|netcoreapp3.1|6.577&nbsp;μs|0.0378&nbsp;μs|0.0315&nbsp;μs|0.0340| 0.0000|0.0000|2.61&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #1839

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SpanBenchmark.StartFinishScope&#8209;netcoreapp3.1 | 1.152 | 1,134.08 | 984.21 | 
| Benchmarks.Trace.SpanBenchmark.StartFinishSpan&#8209;netcoreapp3.1 | 1.145 | 929.77 | 812.15 | 
| Benchmarks.Trace.SpanBenchmark.StartFinishSpan&#8209;net472 | 1.143 | 962.08 | 841.38 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`StartFinishSpan`|net472|961.6&nbsp;ns|10.08&nbsp;ns|8.94&nbsp;ns|0.0681| 0.0000|0.0000|433&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`StartFinishSpan`|netcoreapp3.1|928.9&nbsp;ns|8.86&nbsp;ns|8.29&nbsp;ns|0.0060| 0.0000|0.0000|432&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`StartFinishScope`|net472|1,160.4&nbsp;ns|16.29&nbsp;ns|14.44&nbsp;ns|0.0804| 0.0000|0.0000|514&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/976bf737f468e0d213bc73a423e5610b9923000e)|`StartFinishScope`|netcoreapp3.1|1,133.9&nbsp;ns|10.08&nbsp;ns|9.43&nbsp;ns|0.0078| 0.0000|0.0000|552&nbsp;B|
|#1839|`StartFinishSpan`|net472|845.5&nbsp;ns|13.86&nbsp;ns|12.97&nbsp;ns|0.0679| 0.0000|0.0000|433&nbsp;B|
|#1839|`StartFinishSpan`|netcoreapp3.1|821.3&nbsp;ns|16.27&nbsp;ns|16.71&nbsp;ns|0.0061| 0.0000|0.0000|432&nbsp;B|
|#1839|`StartFinishScope`|net472|1,121.7&nbsp;ns|20.11&nbsp;ns|17.83&nbsp;ns|0.0807| 0.0000|0.0000|514&nbsp;B|
|#1839|`StartFinishScope`|netcoreapp3.1|989.3&nbsp;ns|19.52&nbsp;ns|20.89&nbsp;ns|0.0078| 0.0000|0.0000|552&nbsp;B|
</details>

kevingosse reviewed (approved):
LGTM

andrewlock commented:
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c38ccce959d112e730be13cbf6aa3e5b728dcbd9) will **not change** line coverage
:heavy_check_mark: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c38ccce959d112e730be13cbf6aa3e5b728dcbd9) will **not change** branch coverage
:no_entry: Merging #1839 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c38ccce959d112e730be13cbf6aa3e5b728dcbd9) will will **increase** complexity by `4`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c38ccce959d112e730be13cbf6aa3e5b728dcbd9) | #1839       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9458` / `15198` | `9480` / `15193` |          |
| Lines %   | `62%`      | `62%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4151` / `6590` | `4157` / `6596` |          |
| Branches %| `63%`      | `63%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7766`      | `7770`        |  `4`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=77066&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/77066/artifacts?artifactName=Code%20Coverage%20Report_77066&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1839](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76985&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76985/artifacts?artifactName=Code%20Coverage%20Report_76985&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c38ccce959d112e730be13cbf6aa3e5b728dcbd9) | #1839       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `62%`      | `62%`       |  `0%` :heavy_check_mark:  |
| Branches %| `63%`      | `63%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7766`      | `7770`       |  `4` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Ci.Agent.CIAgentWriter](https://github.com/DataDog/dd-trace-dotnet/tree/d1d2035b1dc5ffb1ecb7afde740c9797355300ab/tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs) | `-11%` :no_entry: | `-12%` :no_entry:   | `2` :no_entry: |
| [Datadog.Trace.ClrProfiler.Helpers.AsyncHelper](https://github.com/DataDog/dd-trace-dotnet/tree/d1d2035b1dc5ffb1ecb7afde740c9797355300ab/tracer/src/Datadog.Trace/ClrProfiler/Helpers/AsyncHelper.cs) | `0%` :heavy_check_mark: | `8%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Span](https://github.com/DataDog/dd-trace-dotnet/tree/d1d2035b1dc5ffb1ecb7afde740c9797355300ab/tracer/src/Datadog.Trace/Span.cs) | `13%` :heavy_check_mark: | `1%` :heavy_check_mark:   | `6` :no_entry: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=77066&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/77066/artifacts?artifactName=Code%20Coverage%20Report_77066&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1839](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=76985&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/76985/artifacts?artifactName=Code%20Coverage%20Report_76985&api-version=6.0&%24format=zip)
